### ISSUE TITLE
assertion failed when reinterpreting slice len

### ISSUE THREAD

**Issue by Vexu** [ziglang/zig#12937]
```zig
pub export fn entry() void {
    comptime var fmt = "foo{}bar{}\n";
    comptime var cleaned: [fmt.len]u8 = undefined;
    comptime var cleaned_i = 0;
    comptime var i = 0;
    inline while (i < fmt.len) {
        const start_index = i;
        inline while (i < fmt.len) : (i += 1) {
            switch (fmt[i]) {
                '{', '}' => break,
                else => {},
            }
        }
        comptime var end_index = i;

        if (start_index != end_index) {
            const str = fmt[start_index..end_index];
            cleaned[cleaned_i..][0..str.len].* = str.*;
            cleaned_i += str.len;
        }
        i += 2;
    }
}
```
```
thread 36134 panic: reached unreachable code
Analyzing a.zig: a.zig:entry
      %97 = dbg_block_begin())
      %98 = dbg_block_begin())
      %99 = dbg_stmt(17, 13)
      %100 = load(%86) 
      %101 = slice_end(%4, %40, %100) 
      %102 = dbg_var_val(%101, "str")
      %103 = dbg_stmt(18, 13)
      %104 = load(%17) 
      %105 = slice_start(%13, %104) 
      %106 = ref(%105) 
      %107 = field_val(%101, "len") 
      %108 = slice_end(%106, @Zir.Inst.Ref.zero, %107) 
      %109 = validate_deref(%108) 
      %110 = validate_deref(%101) 
      %111 = load(%101) 
    > %112 = store_node(%108, %111) 
      %113 = dbg_stmt(19, 13)
      %114 = load(%17) 
      %115 = typeof(%114) 
      %116 = field_val(%101, "len") 
      %117 = add(%114, %116) 
      %119 = dbg_block_end())
      %120 = dbg_block_end())
      %118 = store(%17, %117)
      %121 = break(%96, @Zir.Inst.Ref.void_value)
    For full context, use the command
      zig ast-check -t a.zig

  in a.zig: a.zig:entry
    > %95 = condbr(%94, {%97..%121}, {%122}) 
  in a.zig: a.zig:entry
    > %96 = block({%92..%95}) 
  in a.zig: a.zig:entry
    > %33 = condbr_inline(%32, {%37..%133}, {%134}) 
  in a.zig: a.zig:entry
    > %34 = block_inline({%27..%33}) 
  in a.zig: a.zig:entry
    > %26 = block_inline({%34, %36}) 

./lib/std/debug.zig:281:14: 0x8588b3 in assert (zig)
    if (!ok) unreachable; // assertion failure
             ^
./lib/std/math/big/int.zig:2211:15: 0xe08864 in writeTwosComplement (zig)
        assert(buffer.len >= byte_count);
              ^
./src/value.zig:1228:90: 0x12035f0 in writeToMemory (zig)
                bigint.writeTwosComplement(buffer, bits, abi_size, target.cpu.arch.endian());
                                                                                         ^
./src/value.zig:1256:34: 0x120449c in writeToMemory (zig)
                    writeToMemory(elem_val, elem_ty, mod, buffer[buf_off..]);
                                 ^
./src/Sema.zig:24355:38: 0x11fcd0e in storePtrVal (zig)
            operand_val.writeToMemory(operand_ty, sema.mod, buffer[reinterpret.byte_offset..]);
                                     ^
./src/Sema.zig:24258:33: 0xf8d458 in storePtr2 (zig)
            try sema.storePtrVal(block, src, ptr_val, operand_val, elem_ty);
                                ^
./src/Sema.zig:4499:67: 0x11e2ec3 in zirStoreNode (zig)
    return sema.storePtr2(block, src, ptr, src, operand, src, if (is_ret) .ret_ptr else .store);
                                                                  ^
./src/Sema.zig:1049:38: 0xf80dcb in analyzeBodyInner (zig)
                try sema.zirStoreNode(block, inst);
                                     ^
```

**Comment by Vexu** [comment]
Reduction:
```zig
comptime {
    var arr: [10]u8 = undefined;
    var index: u32 = 1;
    arr[index..][0..2].* = foo();
}
fn foo() [2]u8 {
    return "fooo"[0..2].*;
}
```

**Comment by mlugg** [comment]
Was resolved by 1e6d7f7

**Comment by Vexu** [comment]
That fixed the reduction and the original is now a duplicate of #14158
